pennylane/compiler.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This subpackage provides the JIT compilation support for PennyLane.
+
+
+This subpackage exclusively functions as a wrapper for PennyLane's JIT compiler packages, without independently
+implementing any compiler itself. Currently, it supports the ``pennylane-catalyst`` package, with plans to
+incorporate additional packages in the near future.
+
+For any compiler packages seeking to be registered, it is imperative that they expose the 'entry_points' metadata
+under the designated group name: ``pennylane.compilers``.
+"""
+
+from typing import List
+from importlib import reload
+import pkg_resources
+
+
+class Compiler:
+    """The JIT compiler module containing all utilities and wrapper methods for ``qml.qjit`` in PennyLane."""
+
+    # Private class properties
+    __BACKENDS = []
+    __ENTRY_POINTS = {}
+
+    @classmethod
+    def available(cls, name: str = "pennylane-catalyst") -> bool:
+        """Check the availability of the given compiler package.
+
+        Args:
+            name (str): the name of the compiler package
+
+        Return:
+            bool : True if the compiler package is installed on the system
+        """
+
+        reload(pkg_resources)
+        pkg_name = pkg_resources.Requirement.parse(name)
+
+        if pkg_resources.working_set.find(pkg_name):
+            if name not in cls.__BACKENDS:
+                cls.__ENTRY_POINTS = {
+                    entry.name: entry
+                    for entry in pkg_resources.iter_entry_points("pennylane.compilers")
+                }
+                cls.__BACKENDS.append(name)
+            return True
+
+        if name in cls.__BACKENDS:
+            cls.__BACKENDS.remove(name)
+
+        return False
+
+    @classmethod
+    def available_backends(cls) -> List[str]:
+        """Return the available compiler packages."""
+        return cls.__BACKENDS
+
+    @classmethod
+    def active(cls) -> bool:
+        """Check whether the caller is inside a QJIT evaluation context.
+
+        Return:
+            bool : True if the caller is inside a QJIT evaluation context
+        """
+        if not cls.__BACKENDS:
+            raise RuntimeError("There is no available compiler package.")
```

Comment:
This might depend on how we actually want to use it, but should we raise an error here? An alternative is to return false, since when there are no compilers available there will also not be any active compilers.

Reply:
One of the reasons for raising a runtime error instead of simply returning False was to better distinguish all the following cases:
- There is an available compiler and this is called during tracing
- There is an available compiler but it's outside the tracing context
- There is no available compiler

This error is also added to ensure developers call `available` (the setter method of `__BACKENDS`) before `active`.

Reply:
Something to keep in mind is that True/False allows developers to do:

```python
if active:
    ...
else:
    ...
    ... # developer can optionally raise an error here if they want based on additional checks
```

Whereas the current behaviour requires:

```python
try:
    if active:
        ...
except RuntimeError as e:
    ...
```

What I recommend @maliasadi is having `active()` return True (within a tracing context and a compiler is available) / False (no available compiler or not in a tracing context) because I think this is the biggest use case.

But we can add additional utilities here to separate 2 and 3 if you think it might be useful. Do you have an example of where it might come up?

Reply:
The main motivation for this function is to allow developers to do something similar to what they already do with JAX/JIT:

```python
if jit_occuring:
    # do this
else:
    # do something else
```

which is a very common pattern we're already using.

My assumption in this case is that `qml.compiler.active()` will:

1. return True if we are in a tracing context, and 
2. a tracing context _cannot be started_ if there are no compilers available.

Is that correct?

Reply:
I agree with @josh146, it is very nice for users to be able to do a simple if/else instead of the ty/expect.

Reply:
:100: Thanks everyone!